### PR TITLE
fix(compiler): one owner for dict canonicalisation; retire the producer-less `$={name}` marker (#1608, #1617)

### LINKED ISSUE #1608: dict last-value-wins canonicalisation exists in three independent copies with no drift gate (registry const_fold, codegen fold_dict_create_cmd, ValueOps)
Found by an independent centralisation audit of PR #1591's owners (verified at `claude/e4-codecs-subst` d5c369ec1). Structural follow-up to the `parse_dict` miscompilation being fixed in-PR on #1591 — this issue tracks the drift-gate gap that let it happen.

The dict canonicalisation rule (first-occurrence key position, last value wins — `SetDictFromAny`, `tmp/tcl9.0.4/generic/tclDictObj.c:589`; `Tcl_DictObjPut` hash overwrite) is independently implemented in:

1. `tcl-syntax` `ValueOps::dict_pairs` — the intended owner (`rust/tcl-syntax/src/value.rs:273/:288`), with 12 cmd-core + 14 VM + 9 runtime call sites.
2. `rust/tcl-registry/src/const_fold.rs:422` `fold_dict_create` — its own copy of the walk (correct today).
3. `rust/tcl-compiler/src/codegen/helpers.rs:465` `fold_dict_create_cmd` — a third copy (correct today).

And it was **absent** from `const_fold.rs:341 parse_dict`, which fed six compile-time folders (`fold_dict_get/exists/size/keys/values/merge`) first-match semantics — a live miscompilation (`dict get {a 1 a 2} a` folded to `1`; both tclsh oracles say `2`), being fixed on #1591.

Three correct copies plus one missing copy is exactly the shape #1427 was filed about in the VM and #1439 was filed about for list rendering: per-site re-implementations with no cross-crate drift gate, where `cargo xtask owner-resolution` validates only the manifest and cannot detect a surface that never calls the owner.

**Fix shape:** either give the registry/codegen folders a shared canonicalise entry point (registry cannot depend on heavyweight crates — check the dependency direction; `tcl-syntax` is likely already reachable), or add a drift gate: a cross-crate parity test that feeds a duplicate-key/odd-shape corpus through all three implementations plus real tclsh and asserts byte identity, so a fourth copy or a diverging edit fails a named test.

Related: #1427 (VM copies), #1439 (same class, list rendering), #1591.

### LINKED ISSUE #1617: compiler: the `$={name}` internal marker's producer/consumer pair is release-blind by construction (mutation survivor M3b on #1615)
Recorded from the #1615 mutation inventory: mutant **M3b** (flipping the `$={name}` compiler-marker arm's close scan) survives the full test corpus, and survives identically on the pre-#1615 tree — pre-existing, not introduced there.

**Mechanism:** the `$={name}` escape-hatch marker (the compiler's internal spelling for a braced scalar reference) is produced and consumed by a pair that both use a release-blind rule (`strip_suffix('}')` on the consumer side; see also the #1568 handoff's note that `codegen/values.rs::parse_braced_scalar_ref` uses a *last*-`}` rule while `codegen/helpers.rs`'s `$={name}` arm inside `parse_subst_template` used `find('}')` — a third and fourth variant of the close rule). Because producer and consumer are blind in compatible ways, no current program distinguishes them — which is exactly why the mutant survives: **the seam is unpinned, and a future change to either half can diverge silently.**

This is a latent-inconsistency/drift-gate issue, not an observable bug today. It is distinct from #1604 (free-text scanners over user source): this pair operates on the compiler's own internal marker encoding.

**Fix shape:** either route both halves through one shared helper (so they cannot disagree), or — since the marker names are compiler-generated and could be constrained — document and enforce the invariant that marker names never contain `{`/`}`/`\`, and pin it with a test that a brace-bearing name is never emitted into the marker form. Then the mutant should die by name.

Related: #1568 / #1615 (the six user-visible decoder sites, fixed), #1604.

### PR DESCRIPTION
**Pull request by bitwisecook**
## Summary

Two pre-release drift-gate issues, kept as separate commits because they are separate areas.

### #1617 — `$={name}` is the user's literal text, not a compiler marker

The issue records it as a latent producer/consumer blind spot (mutation survivor M3b). Investigating it turned up something stronger: **the marker has no producer at all**, and its consumer is a live wrong-code path.

Nothing in the tree — back to the Rust rewrite commit `4e403da80` — ever emits `$={name}`. The segmenter re-spells a braced variable word verbatim from source (`segmenter::word_piece`, from #1568), and no other pass writes a `$=` prefix. So every word that reached the decoder came from the user's own source, where `$=` is not a substitution trigger in any supported release (`=` is not a name character; `Tcl_ParseVarName`, `tmp/tcl9.0.4/generic/tclParse.c`):

```tcl
set y hi
puts $={y}
```

tclsh8.6.16 and tclsh9.0.4 both print `$={y}`; our VM printed `hi`. Five of the six routes were wrong — bare command word, `set` value, `switch` subject, list argument, proc body; only the array-key route already read it as a literal.

That producer/consumer blindness is precisely why M3b survived the corpus: no program could tell the two halves apart. So the encoding is removed rather than pinned — `SubstPart::Scalar`, its `parse_subst_template` arm, `values::parse_braced_scalar_ref`, and its five call sites (`cmd_subst` x3, `statements`, `expressions`, `cfg_lower`'s switch subject). Nothing is lost: the form the marker existed to spell, `${a(1)}`, already reaches `parse_simple_var_ref`/`load_var`, which agrees with both oracles (`set {a(1)} S; array set a {1 A}; puts ${a(1)}` prints `A` everywhere — `TclObjLookupVar` parses the parens out of the name at lookup time). A note where the decoder used to be records why a future internal marker needs an out-of-band channel (an IR node or a word flag) rather than a spelling in the `$…` space that user source can collide with.

### #1608 — one owner for dict canonicalisation, plus the cross-crate gate

`tcl_syntax::value::canonical_dict_slots` is now the only implementation of "first-occurrence key position, last value wins" (`SetDictFromAny` over `Tcl_DictObjPut`, `tclDictObj.c:589`). It answers in **slot indices** — one `(key_slot, value_slot)` per surviving key — which is what lets every layer share it: the runtime seam maps them onto value handles, the compile-time folders onto their own strings. `DictCreateCmd`'s argument walk is the same rule, so one function covers both entry points. Keys are borrowed and only `Eq + Hash`, so a byte-keyed value model can bind it too.

Bound by `ValueOps::dict_pairs`, `const_fold::parse_dict`, `const_fold::fold_dict_create`, and `codegen::helpers::fold_dict_create_cmd`.

The drift gate is `rust/tcl-vm/tests/dict_canonicalisation_parity.rs`: one duplicate-key / odd-shape corpus through every binding — the `ValueOps` seam via `Vm`, the registry folds via the public `run_const_fold` path the optimiser uses, and the codegen fold — asserted byte-identical to each other **and** to real tclsh8.6/9.0. `owner-resolution` cannot see this class of drift (it validates the manifest, not whether a surface calls the owner), so the gate is semantic.

The WASM runtime's native dict rep is deliberately **not** routed through the owner: it keeps a live key index across mutation, so it canonicalises incrementally (one `Tcl_DictObjPut` per insert) rather than in one walk. It gets its own leg of the same gate instead (`duplicate_keys_canonicalise_like_the_shared_owner`, same corpus), and the contracts doc says so.

**The gate immediately found a second, live defect** (the #1439 class, in the fold-side renderers rather than the codec). `Tcl_Merge` brace-quotes a leading `#` only in list position 0; the registry's `list_join` and the compiler's `fold_cmd_args` / `fold_dict_create_cmd` mapped the single-element quoter over every element:

| fold | was | tclsh 8.6 / 9.0 |
| --- | --- | --- |
| `[dict keys {a 1 # 2}]` | `a {#}` | `a #` |
| `[list a #]` | `a {#}` | `a #` |
| `[lrepeat 2 #]` | `{#} {#}` | `{#} #` |

Same elements, different *string* — and a fold bakes the string into the program. All three renderers now go through `tcl_syntax::list::join_list`, which applies the rule position by position; that also corrects the `lrange`, `lreverse`, `split`, `concat` and `dict create`/`values`/`merge` folds. Rows pinned in `folded_lists_quote_a_leading_hash_only_in_position_zero`, each byte-exact from tclsh8.6.16 and tclsh9.0.4.

## Validation

- [x] I ran relevant tests/checks locally and listed the exact commands.

All runs prefixed with `CARGO_INCREMENTAL=0 CARGO_PROFILE_DEV_DEBUG=0 CARGO_BUILD_JOBS=2`, and with `TCL_LSP_TCLSH86=…/tmp/tcl8616-install/bin/tclsh8.6` exported so the oracle legs actually execute (tclsh9.0 from PATH).

Gates:

- `cargo fmt --all --check` — clean in `rust/` and in `runtime/rust`
- `cargo clippy -p tcl-syntax --all-targets --features num-bigint`, `cargo clippy -p tcl-registry -p tcl-compiler -p tcl-vm --all-targets`, and `cargo clippy --all-targets` in `runtime/rust` — no new warnings (the runtime's pre-existing `needless_borrow` warnings are all in `cmd_fs.rs`, untouched here; `clippy -p tcl-syntax` without `--features num-bigint` fails to build `number_tower.rs` on `rust` already)
- `cargo check --workspace` — green
- `cargo xtask owner-resolution` — OK, 21 rows
- `make lsp-server-wasm-test` — 16/16 checks passed (this change is below `tcl-lsp-server`; no `std::time` use added)

Suites (all green): `tcl-syntax` (with `num-bigint`), `tcl-registry`, `tcl-compiler`, `tcl-vm` (41 suites), `tcl-lsp-core`, `tcl-cli`, `tcl-lsp-server` (1505 e2e), `tcl-spec-studio` (incl. the `reference_doc` drift test), `tcl-irules`, `tcl-spectcl`, `tcl-mcp`, `tcl-explorer`, `tcl-debugger`, `tcl-diagram`, `tcl-cli-support`, `tcl-cmd-core`, `tcl-lsp-db`, `f5-cli`, `f5-xc`, `tcl-bigip`, `tcl-bigip-query`, `tcl-pkg`, `tcl-irule-test`, `tcl-bytecode`, `tcl-dialect`, `tcl-engine-tclvm`, `tcl-spec-hooks`, and `runtime/rust` (457 tests).

One flake seen, **not** a regression: `semantic_tokens_reference_client::large_file_range_semantic_tokens_converges_via_refresh` failed once under a loaded parallel run (server returned the coarse tier for one index) and passed in isolation and in a full clean re-run of `-p tcl-lsp-server --test e2e` (1505 passed).

Mutation verification (each from a green baseline; every mutant reverted, working tree clean):

| mutant | killed by |
| --- | --- |
| revive the `$={…}` decode in `emit_value` | `compiled_dollar_equals_word_is_literal_at_every_release`, `compiled_dollar_equals_word_matches_real_tclsh`, `dollar_equals_word_is_a_literal_not_a_variable_load` |
| revive it in `emit_cmd_subst_arg` | `cmd_subst_arg_dollar_equals_stays_literal`, `emit_cmd_subst_arg_composite_and_special_forms` |
| `canonical_dict_slots`: first value wins | `canonical_dict_slots_is_first_position_last_value`, `dict_canonicalisation_matches_real_tclsh` |
| `canonical_dict_slots`: duplicate moves to the end | same two |
| `list_join`: back to per-element quoting | `folded_lists_quote_a_leading_hash_only_in_position_zero`, and three of the four parity tests |
| a fourth copy of the walk re-added in `fold_dict_create_cmd` (first value wins) | `dict_create_folds_agree_across_registry_and_codegen`, `dict_canonicalisation_matches_real_tclsh` |

Worth noting from that table: the two cross-layer parity tests pass under a mutation of the **owner** itself, because every layer drifts together. That is exactly why the tclsh leg exists, and it is the leg that fails.

## Compiler / diagnostics checklist

- [x] Did this change alter a compiler fact contract? Yes — the dicts owner row gains a public entry point.
- [x] If yes, I updated the owning design doc: `docs/design/contracts/shared-utility-contracts-rust.md` — `canonical_dict_slots` added to the dicts manifest row, with the new gate and the WASM runtime's position described in the named-test section (`cargo xtask owner-resolution` green).
- [x] No diagnostic or optimisation code added or changed, so no `docs/kcs/codes/` page moved.
- [x] No new design doc.

Fixes #1608
Fixes #1617

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01HqRAya5Xmdo4NJSWPm5C8o

---
_Generated by [Claude Code](https://claude.ai/code/session_01HqRAya5Xmdo4NJSWPm5C8o)_

### CODE REVIEW

File: rust/tcl-compiler/src/codegen/helpers.rs [hunk 1]
```diff
@@ -261,17 +265,7 @@ pub fn parse_subst_template(
...
-                // this name as a scalar", but its name is spelt by the same
-                // brace form, so it closes by the same release rule.
-                let end = match tcl_lexer::braced_var_name_end(bytes, i + 2, braced_var) {
-                    tcl_lexer::BracedVarEnd::Closed(end) => end,
-                    tcl_lexer::BracedVarEnd::Unterminated => return None,
-                };
-                parts.push(SubstPart::Scalar(template[i + 2..end].to_owned()));
-                i = end + 1;
-            } else if bytes[i] == b'{' {
+            if bytes[i] == b'{' {
```
[the remark this answers is not in the document]
Comment (bitwisecook) [reply rust/tcl-compiler/src/codegen/helpers.rs]: Confirmed and fixed in 82f956576 (branch `claude/f9-fixes-1608-1617`; doc-lint follow-up d66b82aeb). Thank you — the repro is exact, and the shape is wider than the marker.

**Oracle evidence first** (tclsh8.6.16 via `TCL_LSP_TCLSH86`, tclsh9.0.4 from PATH — the two agree on every row):

| script | tclsh 8.6 / 9.0 | ours before |
| --- | --- | --- |
| `puts [list $={y}$x]` | `{$={y}X}` | `{$={y}$x}` |
| `puts [list a$={y}$x]` | `{a$={y}X}` | `{a$={y}$x}` |
| `puts [string length $={y}$x]` | `6` | `7` |
| `set (k) EMPTY; puts [list a$(k)b$x]` | `aEMPTYbX` | `{a$(k)b$x}` |

Your diagnosis of the mechanism is right, and the reason my #1617 vectors missed it is worth recording: a *top-level* word reaches codegen through the segmenter, which re-spells a bare `$x` as `${x}`, and the runtime `subst_word` fallback resolves that even from a word pushed raw — so `puts $={y}$x`, the `set` value, the `switch` subject and the array key were all already correct. Inside an inline command substitution the argument text is the **raw source**, bare `$x` and all, so there the bail-out lost the substitution outright. My vectors were all lone-`$=` shapes and never mixed a literal `$` with a real substitution in one word.

For the record it is not a #1617 regression: before the marker removal the same word decoded as a scalar load of `y`. Both readings disagree with C; one of them just disagreed more quietly.

**Fixed as you proposed, with the C rule cited in the code.** `Tcl_ParseVarName` documents three forms at `tmp/tcl9.0.4/generic/tclParse.c:1367-1381`; form 3 is "the $ sign is followed by something that isn't a letter, digit, or underscore: there is no variable name and the token is just `$`". C jumps to `justADollarSign` (`:1454` when the name scan consumed nothing, `:1360` for a trailing `$`), rewrites the token as literal text and returns `TCL_OK` — parsing continues. So a `$` that starts no substitution now joins the literal run and the scan carries on.

The `$` reading moved into one `scan_dollar` returning Subst / Literal / Unterminated. That keeps the three forms visible in one place and preserves the one case that really is a bail-out: `${` with no close is C's `TCL_PARSE_MISSING_VAR_BRACE` ("missing close-brace for variable name", `:1414`), so it still declines and the existing unterminated-brace tests stay green.

One thing your repro pulled in for free: form 2's **empty array name**. `$(k)` is a reference to the empty-named array — C admits it explicitly ("Support for empty array names here", `:1449-1453`) — so an empty name still counts when a `(` follows. That composite (`a$(k)b$x`) was un-decodable before and is now correct; only the whole-word `$(k)` had worked, via the other decoder.

**Pinned and mutation-verified.** The bare-`$` unit test now covers the literal reading plus `$x$`, `$ $x`, the mixed `$={y}$x` / `a$={y}$x` / `$={y}${x}` shapes and the empty-name array; nine VM vectors join the #1617 table in `cross_version_vars_e2e` — your repro, the measured string, and the mixed shape on every route the lone-`$=` vectors covered — each compared byte-for-byte with real tclsh8.6 and tclsh9.0 at 8.4/8.5/8.6/9.0/9.1. Restoring the `None` bail-out kills four named tests (`subst_template_bare_dollar_is_literal_text`, `subst_template_dollar_equals_is_not_a_marker`, `compiled_dollar_equals_word_is_literal_at_every_release`, `compiled_dollar_equals_word_matches_real_tclsh`); dropping the empty-array-name case kills `subst_template_empty_array_name_is_a_reference` plus the two VM tests. The earlier six-mutant battery (two marker-revive, four dict) was re-run on top of this fix and all six still die by name; every mutant reverted, tree clean.

Gates re-run green: `cargo test -p tcl-compiler` (63 suites), `-p tcl-vm` (41 suites), `cargo fmt --all --check`, `cargo clippy -p tcl-compiler -p tcl-vm --all-targets`, `cargo check --workspace`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01HqRAya5Xmdo4NJSWPm5C8o

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by bitwisecook** [comment]
Lane F9 checkpoint — branch `claude/f9-fixes-1608-1617`, head d66b82aeb.

Addressed the P2 on thread r3824340877 (reply with the full oracle evidence is on the thread itself):

- **82f956576** — `fix(compiler): a $ that starts no substitution is literal text, not a bail-out`. `parse_subst_template` gave up on the whole word when a `$` was not followed by a name; inside an inline command substitution the argument text is the raw source, so the word was pushed raw and a real `$x` later in it was never substituted. `[list $={y}$x]` gave `{$={y}$x}` where both oracles give `{$={y}X}`. Now follows `Tcl_ParseVarName`'s third form (`justADollarSign`, `tmp/tcl9.0.4/generic/tclParse.c:1454` / `:1502`): the `$` is text and the scan continues. `${` with no close still declines, because C makes that an error (`:1414`). Form 2's empty array name (`$(k)`, `:1449-1453`) is kept as a reference, which also fixes composites like `a$(k)b$x`.
- **d66b82aeb** — doc-lint follow-up (unbalanced backtick in the new doc comment).

New coverage: the bare-`$` unit test plus nine VM vectors in `cross_version_vars_e2e`, all byte-compared against real tclsh8.6/9.0 at 8.4/8.5/8.6/9.0/9.1. Two new mutants (restore the bail-out; drop the empty-array-name case) die by name, and the earlier six-mutant battery was re-run on top and still dies. Gates green again: `cargo test -p tcl-compiler` (63 suites), `-p tcl-vm` (41 suites), `fmt --all --check`, `clippy -p tcl-compiler -p tcl-vm --all-targets`, `cargo check --workspace`.

Not chasing the red pr-gate — that is the repo-wide Rust 1.98 lint breakage F10 is fixing, not this branch.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by bitwisecook** [comment]
Conflict cleared — branch `claude/f9-fixes-1608-1617`, head **cbc127e8f** (merge of origin/rust 8f2719c99, #1669's Rust 1.98 clippy sweep).

Three conflicts, all the same shape: #1669 had rewritten the per-site duplicate-key walks (`chunks_exact(2)` → `as_chunks::<2>().0`) in `ValueOps::dict_pairs`, `const_fold::parse_dict`, `const_fold::fold_dict_create` and `codegen::helpers::fold_dict_create_cmd` — which are exactly the copies this PR deletes. Resolved by keeping the centralised form (the lint sites go away with the loops) and taking `as_chunks::<2>().0` in the one place my code still chunks, the key extraction in `dict_pairs`. No semantic change from either side was dropped.

Gates re-run on the merge result, all green:

- `cargo fmt --all --check` (rust/ and runtime/rust)
- clippy under **1.98**: `-p tcl-syntax --all-targets --features num-bigint`, `-p tcl-registry -p tcl-compiler -p tcl-vm --all-targets` — no warnings, so no new lint sites in the added code
- `cargo check --workspace`
- `cargo test -p tcl-compiler` (63 suites); `-p tcl-vm -p tcl-registry -p tcl-syntax` (65 suites, incl. `cross_version_vars_e2e` 27/27 and `dict_canonicalisation_parity` 4/4 with tclsh8.6.16 + tclsh9.0.4 live); `-p tcl-lsp-core -p tcl-spec-studio` (45 suites, `reference_doc` drift test green); `runtime/rust` dict tests
- `make lsp-server-wasm-test` — 16/16

The Codex P2 (thread r3824340877) is answered on the thread and fixed in 82f956576; I could not mark the thread resolved from here — the GraphQL review-threads read is rate-limited on this token, so I have no thread node ID. Happy for you to resolve it, or I can retry later.

Ready for merge when you are.

---
_Generated by [Claude Code](https://claude.ai/code)_
